pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+			return err
+		}
+	}
+
+	log.Infof("VerifyCloudCertChain: key count %d\n", keyCnt)
+	if leafcert == nil {
+		err := fmt.Errorf("VerifyCloudCertChain: fail to acquire leaf cert")
+		return err
+	}
+
+	// Get the rootCAs from the zedcloudCtx
+	var roots *x509.CertPool
+	if ctx.TlsConfig != nil {
+		roots = ctx.TlsConfig.RootCAs
+	} else {
+		err := fmt.Errorf("VerifyCloudCertChain: no rootCAs in TlsConfig")
+		return err
+	}
+
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		DNSName:       serverName,
+		Intermediates: interm,
+	}
+	if _, err := leafcert.Verify(opts); err != nil {
+		log.Infof("VerifyCloudCertChain: fail to verify the cert chain %v\n", err)
+		return err
+	}
+
+	// save this cert and hash of it
+	cloudCertHash = ComputeSha(certByte)
+
+	err = saveServerCertToFile(certByte)
+	if err != nil {
+		log.Infof("VerifyCloudCertChain: file save err %v\n", err)
+		return err
+	}
+	log.Infof("VerifyCloudCertChain: success\n")
+	return nil
+}
+
+// saveCertsToFile - save cloud signing certs into files
+func saveServerCertToFile(certByte []byte) error {
+	tmpfileName := types.ServerCertName + ".tmp"
+	pemTmpCertFile, err := os.Create(tmpfileName)
+	if err != nil {
+		log.Infof("SaveCertsToFile: create tmp file err %v\n", err)
+		return err
+	}
+
+	_, err = pemTmpCertFile.Write(certByte)
+	if err != nil {
+		log.Infof("SaveCertsToFile: write pem to file err %v\n", err)
+		return err
+	}
+
+	defer os.Remove(pemTmpCertFile.Name())
+	pemTmpCertFile.Sync()
+	if err := pemTmpCertFile.Close(); err != nil {
+		log.Errorln("SaveCertsToFile: ", tmpfileName, err)
+		return err
+	}
+
+	if err := os.Rename(tmpfileName, types.ServerCertName); err != nil {
+		log.Errorln("SaveCertsToFile: ", err)
+		return err
+	}
+	log.Infof("SaveCertsToFile: saved cert to %s\n", types.ServerCertName)
+	return nil
+}
+
+// RemoveCertFile - remove the server certificate file
+func RemoveCertFile() {
+	_, err := os.Stat(types.ServerCertName)
+	if err == nil {
+		err := os.Remove(types.ServerCertName)
+		if err != nil {
+			log.Errorln("RemoveCertFile: ", err)
+		}
+	}
+}
+
+// UseV2API - check the controller cert file and use V2 api if it exist
+func UseV2API() bool {
+	_, err := os.Stat(types.ServerCertName)
+	if err == nil {
+		return true
+	}
+	return false
+}
+
+// URLPathString - generate url for either v1 or v1 API path
+func URLPathString(server string, isV2api, isHTTP bool, devUUID uuid.UUID, action string) string {
+	var urlstr string
+	if !isV2api {
+		urlstr = server + "/api/v1/edgedevice/" + action
+	} else {
+		urlstr = server + "/api/v2/edgedevice/"
+		if devUUID != nilUUID {
+			urlstr = urlstr + "id/" + devUUID.String() + "/"
+		}
+		urlstr = urlstr + action
+	}
+	if isHTTP {
+		return "http://" + urlstr
+	}
+	return urlstr
+}
+
+// the cloud controller lookup prefers the hash computed from x509 cert
+// also need the CN for onboarding when register
+func certToSha256(cert tls.Certificate, onboard bool) ([]byte, error) {
```

Comment:
onboard argument is no longer used hence can be removed.
But don't we already store this sha in a global variable?

Reply:
ok. missed that.